### DISCUSSION TITLE
New event on user updates by IdP Mappers

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#19493] [Ideas]
When a user changes values at a brokered IdP (e.g. last name) a mapper ("Attribute Importer") can be used to force an update of a corresponding user attribute at Keycloak. However there is no event created at Keycloak for the update operation.  This makes it hard to keep track of user updates, based on events. Basically all login events need to be considered as potential updates and it need to be checked whether the user attributes did actually change.

**Proposal**
In [IdentityBrokerService](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L990) at method `updateFederatedIdentity` all IdP mappers are executed. Also the basic user attributes may be changed in this method (see `setBasicUserAttributes`).

The user attributes after execution could be compared to the attributes before execution. If setBasicUserAttributes or a mapper updated any attribute value, a new login event could be created (e.g. IDENTITY_PROVIDER_USER_UPDATE). The event should have all updated fields and values in the details.

WDYT? Any other ideas to get an event in case of user updates by IdP Mappers?

**Comment by Romain7495** [comment]
Hello, 

I totally agree, it’s hard to track identity provider User Profile updates as no event is triggered when the Attribute Importer forces the update of a User Profile field.

Can we get some direction from a maintainer, please? I can try a PR if we have some help to start.

Maybe @pedroigor @ahus1

Help is really much appreciated!

**Reply by ahus1** [reply]
I agree such an event would be helpful. I'd say a combined event that mentions that the user or some of attribute has been updated should be sufficient unless there is an example that demands otherwise. 

The current Keycloak events don't capture all the data that has been changed, to "The event should have all updated fields and values in the details." would IMHO go beyond what we usually put into events. With the changes happening here, it might become difficult to ensure "The user attributes after execution could be compared to the attributes before execution. " ... instead, I'd say whenever `updateFederatedIdentity` is called (which is usually wrapped in a check if it needs to be called) or `setDiffAttrToConsumer` (which already checks for a change in the attribute) is called, and event should be created.

Note that an updated token leads to an updated user today, which would lead to an event.
